### PR TITLE
feat: Move My Expenses to Platform: Fixing Select Functionality and Create Report From List

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖 Generated by Copilot at 3b31029</samp>

This pull request refactors some components to use the new `PlatformExpense` model and the `sharedExpenseService` for working with expenses from the platform API. It also adds a new method to the `ExpensesService` to fetch all expenses without pagination. The purpose of these changes is to migrate from the deprecated transaction API and improve code quality and maintainability.

### <samp>🤖 Generated by Copilot at 3b31029</samp>

> _Sing, O Muse, of the skillful coder who refactored the code_
> _To use the new `PlatformExpense` model and the `ExpensesService`_
> _And who created the `sharedExpenseService` to avoid duplication_
> _And to make the components work with harmony and grace._

### <samp>🤖 Generated by Copilot at 3b31029</samp>

*  Add a new file `shared-expense.service.ts` that contains common methods for handling expenses across different components ([link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8a83a46f72dc182a7e19da6a6985ac4ffaee1d56ee70c0d94165d17e0a2cff98R26-R103))
*  Add a new method `getAllExpenses` to the `ExpensesService` class that fetches all the expenses from the API based on the given query parameters ([link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-43a6608834bfbe4963726f4de34dd6082b69581858020acfa23be1e61037d9f5R37-R54))
*  Import and inject the `SharedExpenseService` in the `MyExpensesPage` component and use its methods instead of the deprecated `TransactionService` ([link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533R84), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L224-R226), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L920-R923), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L993-R998), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1050-R1053), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1224-R1228), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1280-R1292), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1292-R1300), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1344-R1383), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1375-R1389), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1381-R1395))
*  Update the types and properties of the `selectedElements` and `expensesToBeDeleted` properties and the `expense` parameter of the `MyExpensesPage` component from `Partial<Expense>` to `Partial<PlatformExpense>` and use the new model's properties instead of the old ones ([link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L133-R134), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L185-R186), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L261-R263), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L882-R891), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L895-R898), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L903-R911), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L987-R990), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1012-R1015), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1109-R1112), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1268-R1278))
*  Update the logic of the `showNewReportModal`, `deleteAllExpenses`, `showOldReportsMatBottomSheet`, `deleteSelectedExpenses`, `openDeleteExpensesPopover`, `selectAllExpenses`, and `unselectAllExpenses` methods of the `MyExpensesPage` component to use the new services and models that are compatible with the platform API ([link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L987-R990), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L993-R998), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1012-R1015), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1050-R1053), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1109-R1112), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1224-R1228), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1268-R1278), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1280-R1292), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1292-R1300), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1300-R1311), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1344-R1383), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1375-R1389), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1381-R1395))
*  Update the imports in the `CreateNewReportComponent` component and use the new `PlatformExpense` model instead of the old `Expense` model ([link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L4-R6))
*  Update the logic of the `getReportTitle`, `ionViewWillEnter`, `selectExpense`, and `ctaClickedEvent` methods of the `CreateNewReportComponent` component to use the new model's properties instead of the old ones ([link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L52-R55), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L78-R85), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L118-R118))
*  Add the `void` or `Subscription` return types for the methods of the `CreateNewReportComponent` component as per the best practices ([link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L52-R55), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L66-R66), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L78-R85), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L93-R93), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L102-R106), [link](https://github.com/fylein/fyle-mobile-app/pull/2557/files?diff=unified&w=0#diff-4ad3a3445398cc6db66f3e3f9d099b3b3d6545ad52c7fd271c1be6401c87aed7L118-R118))

## Clickup
https://app.clickup.com/t/86ctv3kk4

## Code Coverage
Please add code coverage here

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

**Review by jayfyle — COMMENTED** [review]
The selecting and delete is commented out for now, fixing it in upcoming PRs

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 1]
```diff
@@ -32,4 +32,78 @@ export class ExpenseService {
...
+  }
+
+  excludeCCCExpenses(expenses: Partial<Expense>[]): Partial<Expense>[] {
+    return expenses.filter((expense) => expense && expense.matched_corporate_card_transaction_ids.length > 0);
+  }
+
+  isMergeAllowed(expenses: Partial<Expense>[]): boolean {
+    if (expenses.length === 2) {
+      const areSomeMileageOrPerDiemExpenses = expenses.some(
+        (expense) => expense.category?.system_category === 'Mileage' || expense.category?.system_category === 'Per Diem'
```
Comment (Chethan-Fyle) [inline src/app/core/services/platform/v1/shared/expense.service.ts:49]: `(expense) =>  ['Mileage', 'Per Diem'].includes(expense.category?.system_category)`

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 2]
```diff
@@ -32,4 +32,78 @@ export class ExpenseService {
...
   }
+
+  getReportableExpenses(expenses: Partial<Expense>[]): Partial<Expense>[] {
+    return expenses.filter(
+      (expense) => !this.getIsCriticalPolicyViolated(expense) && !this.getIsDraft(expense) && expense.id
+    );
+  }
+
+  excludeCCCExpenses(expenses: Partial<Expense>[]): Partial<Expense>[] {
+    return expenses.filter((expense) => expense && expense.matched_corporate_card_transaction_ids.length > 0);
```
Comment (Chethan-Fyle) [inline src/app/core/services/platform/v1/shared/expense.service.ts:43]: `return expenses.filter((expense) => expense.matched_corporate_card_transaction_ids?.length > 0);`

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 3]
```diff
@@ -32,4 +32,78 @@ export class ExpenseService {
...
+  }
+
+  isMergeAllowed(expenses: Partial<Expense>[]): boolean {
+    if (expenses.length === 2) {
+      const areSomeMileageOrPerDiemExpenses = expenses.some(
+        (expense) => expense.category?.system_category === 'Mileage' || expense.category?.system_category === 'Per Diem'
+      );
+      const areAllExpensesSubmitted = expenses.every((expense) =>
+        ['APPROVER_PENDING', 'APPROVED', 'PAYMENT_PENDING', 'PAYMENT_PROCESSING', 'PAID'].includes(expense.state)
```
Comment (Chethan-Fyle) [inline src/app/core/services/platform/v1/shared/expense.service.ts:52]: Don't we have enum for the expense state? Create one if not already present.

File: src/app/shared/components/create-new-report/create-new-report.component.ts [hunk 4]
```diff
@@ -49,10 +49,10 @@ export class CreateNewReportComponent implements OnInit {
     private expenseFieldsService: ExpenseFieldsService
   ) {}
 
-  getReportTitle() {
-    const txnIds = this.selectedElements.map((etxn) => etxn.tx_id);
+  getReportTitle(): Subscription {
+    const txnIds = this.selectedElements.map((etxn) => etxn.id);
     this.selectedTotalAmount = this.selectedElements.reduce(
-      (acc, obj) => acc + (obj.tx_skip_reimbursement ? 0 : obj.tx_amount),
+      (acc, obj) => acc + (obj.is_reimbursable ? 1 : obj.amount),
```
Comment (Chethan-Fyle) [inline src/app/shared/components/create-new-report/create-new-report.component.ts:55]: Why are we adding one to the total amount? 🤔

**Review by Chethan-Fyle — CHANGES_REQUESTED** [review]
Minor comments and a doubt.

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 5]
```diff
@@ -6,11 +6,11 @@ import { Expense } from 'src/app/core/models/platform/v1/expense.model';
   providedIn: 'root',
 })
 export class ExpenseService {
-  getIsDraft(expense: Expense): boolean {
+  getIsDraft(expense: Partial<Expense>): boolean {
```
Comment (OmkarJ13) [inline src/app/core/services/platform/v1/shared/expense.service.ts:9]: Why is partial needed here?

File: src/app/core/services/platform/v1/shared/expense.service.ts (hunk 3 above)
Comment (OmkarJ13) [reply src/app/core/services/platform/v1/shared/expense.service.ts:52]: ➕

File: src/app/core/services/platform/v1/shared/expense.service.ts (hunk 1 above)
Comment (OmkarJ13) [reply src/app/core/services/platform/v1/shared/expense.service.ts:49]: ➕

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 6]
```diff
@@ -32,4 +32,78 @@ export class ExpenseService {
...
+
+  hasCriticalPolicyViolation(expense: Partial<Expense>): boolean {
+    return typeof expense.policy_amount === 'number' && expense.policy_amount < 0.0001;
+  }
+
+  getExpenseDeletionMessage(expensesToBeDeleted: Partial<Expense>[]): string {
+    return `You are about to permanently delete ${
+      expensesToBeDeleted.length === 1 ? '1 selected expense.' : expensesToBeDeleted.length + ' selected expenses.'
+    }`;
+  }
```
Comment (OmkarJ13) [inline src/app/core/services/platform/v1/shared/expense.service.ts:71]: Can we create a pipe for this, basically in html we should be able to specify if we should add `s` at the end based on a condition. This is a common use case and having a pipe would make things much cleaner

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 7]
```diff
@@ -32,4 +32,78 @@ export class ExpenseService {
...
+
+  getCCCExpenseMessage(expensesToBeDeleted: Partial<Expense>[], cccExpenses: number): string {
+    return `There ${cccExpenses > 1 ? 'are' : 'is'} ${cccExpenses} corporate card ${
+      cccExpenses > 1 ? 'expenses' : 'expense'
+    } from the selection which can\'t be deleted. ${
+      expensesToBeDeleted?.length > 0 ? 'However you can delete the other expenses from the selection.' : ''
+    }`;
+  }
```
Comment (OmkarJ13) [inline src/app/core/services/platform/v1/shared/expense.service.ts:79]: Same as above

File: src/app/shared/components/create-new-report/create-new-report.component.ts (hunk 4 above)
Comment (OmkarJ13) [reply src/app/shared/components/create-new-report/create-new-report.component.ts:55]: ➕

**Review by OmkarJ13 — CHANGES_REQUESTED** [review]

File: src/app/shared/components/create-new-report/create-new-report.component.ts (hunk 4 above)
Comment (jayfyle) [reply src/app/shared/components/create-new-report/create-new-report.component.ts:55]: We display the total amount of selected expenses on top of the modal with the number of selected expenses

File: src/app/core/services/platform/v1/shared/expense.service.ts (hunk 6 above)
Comment (jayfyle) [reply src/app/core/services/platform/v1/shared/expense.service.ts:71]: Will create this separately to not make this PR too long.

**Review by OmkarJ13 — APPROVED** [review]

File: src/app/core/services/platform/v1/shared/expense.service.ts (hunk 6 above)
Comment (OmkarJ13) [reply src/app/core/services/platform/v1/shared/expense.service.ts]: sure

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 8]
```diff
@@ -1265,39 +1268,47 @@ export class MyExpensesPage implements OnInit {
...
           this.cccExpenses,
           expenseDeletionMessage,
           cccExpensesMessage
         ),
         ctaText: this.expensesToBeDeleted.length > 0 && this.cccExpenses > 0 ? 'Exclude and Delete' : 'Delete',
         disableDelete: this.expensesToBeDeleted.length > 0 ? false : true,
-        deleteMethod: () => this.deleteSelectedExpenses(offlineExpenses),
+        deleteMethod: () => {
+          this.deleteSelectedExpenses();
+          // this.deleteOutboxExpense(offlineExpenses);
```
Comment (Dimple16) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1310]: let's remove this if unused or add a `TODO`

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 9]
```diff
@@ -1341,44 +1352,47 @@ export class MyExpensesPage implements OnInit {
     if (checked) {
       this.selectedElements = [];
       if (this.pendingTransactions.length > 0) {
-        this.selectedElements = this.pendingTransactions;
+        // this.selectedElements = this.pendingTransactions;
```
Comment (Dimple16) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1355]: same as above

**Review by Dimple16 — APPROVED** [review]
minor comments

File: src/app/fyle/my-expenses/my-expenses.page.ts (hunk 8 above)
Comment (jayfyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1310]: Fixed in later PRs

File: src/app/fyle/my-expenses/my-expenses.page.ts (hunk 9 above)
Comment (jayfyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1355]: Fixed in later PRs

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 10]
```diff
@@ -32,4 +32,84 @@ export class ExpenseService {
 
     return vendorDisplayName;
   }
+
+  getReportableExpenses(expenses: Expense[]): Expense[] {
+    return expenses.filter(
+      (expense) => !this.getIsCriticalPolicyViolated(expense) && !this.getIsDraft(expense) && expense.id
```
Comment (Chethan-Fyle) [inline src/app/core/services/platform/v1/shared/expense.service.ts:38]: `expense.id && !this.getIsCriticalPolicyViolated(expense) && !this.getIsDraft(expense)`
- Put `expense.id` first as there is no need to make two function calls if id itself is not present.

**Review by Chethan-Fyle — APPROVED** [review]

**Pull request merged**